### DISCUSSION TITLE
Updating the current entity's id does not update the current entity

### DISCUSSION THREAD

**Discussion by pBouillon** [state-adapt/state-adapt#74] [Q&A] [answered]
Hi, and thank you for the time you've been putting in answering questions and taking care of StateAdapt!

I am using it on a side project in order to get used to it and discover its API in more depth before potentially trying it with a wider project.

I am currently trying to implement some kind of game, and managing the rounds from my state. The idea is the following:
I have a collection of `Player` entities along with the id of the current one and a selector to compute the current player:

```ts
export const initialState: Readonly<GameState> = {
  players: createEntityState<Player, PlayerId, GameState['players']>(),
  currentPlayerId: 0,
};

export const gameAdapter = joinAdapters<GameState>()({
  players: playerEntityAdapter,
  currentPlayerId: numberAdapter,
})({
  currentPlayer: (state) => state.playersAll.at(state.currentPlayerId),
})();
```

In my `GameService`, I rely on the `GameState` to handle the end of the turn by exposing a `endCurrentRound$` source.
I then have a `#roundEnded$` observable that derives from this source to compute the id of the next player and update it:

```ts
@Injectable({ providedIn: 'root' })
export class GameService {
  // Sources
  readonly endCurrentRound$ = new Source<void>('[Game] End Current Round');
  readonly #setNextPlayerId$ = new Source<number>('[Game] Set Next Player Id');

  // State
  readonly #store = adapt(initialState, {
    path: 'game',
    adapter: gameAdapter,
    sources: {
      setCurrentPlayerId: this.#setNextPlayerId$,
    },
  });

  // Selectors
  readonly currentPlayer = toSignal<Player | null>(this.#store.currentPlayer$);

  // Effects
  readonly #roundEnded$ = this.endCurrentRound$.pipe(
    withLatestFrom(this.#store.currentPlayer$),
    map(([, player]: [Action<void>, Player]) => player.id),
    withLatestFrom(this.#store.playersIds$),
    map(([playerId, playerIds]: [string, string[]]) => {
      const playerIdIndex = playerIds.indexOf(playerId);
      const nextPlayerIdIndex = (playerIdIndex + 1) % playerIds.length;
      return Number(playerIds[nextPlayerIdIndex]);
    }),
    tap((playerId) => this.#setNextPlayerId$.next(playerId)),
  );
}
```

From my component, I simply call the source:

```ts
readonly #gameService = inject(GameService);

endCurrentRound(): void {
  this.#gameService.endCurrentRound$.next();
}
```

I was expecting my `currentPlayer` selector to reflect that change, but unfortunately it doesn't.

After observing the Redux devtools, it seems that the `[Game] End Current Round` event is never fired from the `endCurrentRound$` source.

I was wondering:

- Did I missed something in the way of calling my initial trigger?
- Do I need to subscribe to something more than `this.#store.currentPlayer$` to have my changes properly reflected?
- Is this the right way of composing side-effects from my trigger? The next step will be to pass the round details to the `endCurrentRound$` source, as a `Source<Round>`, and having two effects depending on it: one to add the round to the history and another one to compute the id of the current player, is this approach correct or is there another mor idiomatic way of representing that with StateAdapt?

Thanks a lot for your time, your patience and the efforts you are investing in this library!

**Comment by mfp22** [comment, marked as answer]
Thank you. I wish I had more time, but I think I can do more than I have recently. 

You aren't subscribing to `roundEnded$` it looks like. And StateAdapt doesn't really separate effects and sources like other libraries, because most state management libraries handle effects in an imperative way. Here you have an imperative `tap((playerId) => this.#setNextPlayerId$.next(playerId)),` so StateAdapt provides an alternate way to handle what you're doing. 

The store should declare its own dependency on `roundEnded$`. To get around the circular reference issue, you can use `watch` from `@state-adapt/angular`:

```typescript
@Injectable({ providedIn: 'root' })
export class GameService {
  readonly path = 'game';

  // Sources
  readonly endCurrentRound$ = new Source<void>('[Game] End Current Round');
  readonly #roundEnded$ = this.endCurrentRound$.pipe(
    withLatestFrom(watch(this.path, gameAdapter).currentPlayer$),
    map(([, player]: [Action<void>, Player]) => player.id),
    withLatestFrom(watch(this.path, gameAdapter).playersIds$),
    map(([playerId, playerIds]: [string, string[]]) => {
      const playerIdIndex = playerIds.indexOf(playerId);
      const nextPlayerIdIndex = (playerIdIndex + 1) % playerIds.length;
      return Number(playerIds[nextPlayerIdIndex]);
    }),
    toSource('roundEnded$'),
  );

  // State
  readonly #store = adapt(initialState, {
    path: 'game',
    adapter: gameAdapter,
    sources: {
      setCurrentPlayerId: this.#roundEnded$,
    },
  });

  // Selectors
  readonly currentPlayer = toSignal<Player | null>(this.#store.currentPlayer$);
}
```

Also, doesn't `withLatestFrom` allow multiple arguments? That would simplify the beginning of that. I don't remember, but I hope it does. You could probably also create a selector to return what you need and keep the sources code focused on side-effect stuff. Actually with a selector you could get rid of the `withLatestFrom` as well. Yeah, the sources code would be like

```typescript
  // Sources
  readonly endCurrentRound$ = new Source<void>('[Game] End Current Round');
  readonly #nextPlayerId$ = this.endCurrentRound$.pipe(
    switchMap(() => watch(this.path, gameAdapter).nextPlayerId$),
    toSource('nextPlayerId$'),
  );
```

This would wait for `endCurrentRound$` to emit before calculating `nextPlayerId$`.

**Reply by pBouillon** [reply]
Thanks a lot for your answer!

Just so you know, this code throws an error telling that this cannot be run outside of an injection context:

```ts
  readonly #nextPlayerId$ = this.endCurrentRound$.pipe(
    switchMap(() => watch(this.path, gameAdapter).nextPlayerId$),
    toSource('nextPlayerId$'),
  );
```

I suspect `watch` to rely on `inject` and, since it's called from `switchMap`, it fails while evaluating it.

About my issue, I created the selector in the adapter (and renamed `Id` to `Index` for clarity):

```ts
export const initialState: Readonly<GameState> = {
  players: createEntityState<Player, PlayerId, GameState['players']>(),
  currentPlayerIndex: 0,
};

export const gameAdapter = joinAdapters<GameState>()({
  players: playerEntityAdapter,
  currentPlayerIndex: numberAdapter,
})({
  currentPlayer: ({ playersAll, currentPlayerIndex }) =>playersAll.at(currentPlayerIndex) || null,
  nextPlayerIndex: ({ currentPlayerIndex, playersIds }) => (currentPlayerIndex + 1) % playersIds.length,
})();
```

I then used what you suggested to get rid of the "effects" I had:

```ts
@Injectable({ providedIn: 'root' })
export class GameService {
  readonly #path = 'game';

  // 👇 Workaround for the injection context issue
  readonly #watchedAdapter = watch(this.#path, gameAdapter);

  // Sources
  readonly endCurrentRound$ = new Source<void>('[Game] End Current Round');
  readonly #nextPlayerIndex$ = this.endCurrentRound$.pipe(
    switchMap(() => this.#watchedAdapter.nextPlayerIndex$),
    toSource('nextPlayerIndex$'),
  );

  // State
  readonly #store = adapt(initialState, {
    path: this.#path,
    adapter: gameAdapter,
    sources: { setCurrentPlayerIndex: this.#nextPlayerIndex$ },
  });

  // Selectors
  readonly currentPlayer = toSignal<Player | null>(this.#store.currentPlayer$);

  // No more effects 🙌
}
```

While TS doesn't throw any error and is happy with the way this is written, the execution runs into a strange infinite loop when calling `this.#gameService.endCurrentRound$.next();`:

![image](https://github.com/state-adapt/state-adapt/assets/22640284/82044023-bf51-4042-beb8-53fba325f334)

I suspected that being a side effect of the source and the selector having the same name:
![image](https://github.com/state-adapt/state-adapt/assets/22640284/fecd160f-52df-44d0-92ed-8e9b66945bee)

But renaming it results in the same problem, do you know what is going on?

**Reply by mfp22** [reply]
There shouldn't be a naming collision there. 

I suspect the watched observable is emitting over and over again. Does `setCurrentPlayerIndex` cause the returned value to change and re-emit?

**Reply by pBouillon** [reply]
I think that `setCurrentPlayerIndex` is available through the `numberAdapter` it uses, I don't remember it emitting 

However, `nextPlayerIndex` does rely on it (`(currentPlayerIndex + 1) % playersIds.length`), could this trigger the observable's emission? What alternative could I use if the watched adapter has that kind of behavior?

**Reply by mfp22** [reply]
It changes, so the observable emits again. Maybe just a `take(1)` inside the `switchMap` would make sense. Each time the switchMap runs again it would reset and be fine again.

**Reply by pBouillon** [reply]
Wonderful it did work and fixed the issue I had, thanks a lot for your help on this 🙌 

For the record, here is the final code:

```ts
// Adapter
export const initialState: Readonly<GameState> = {
  players: createEntityState<Player, PlayerId, GameState['players']>(),
  currentPlayerIndex: 0,
};

export const gameAdapter = joinAdapters<GameState>()({
  players: playerEntityAdapter,
  currentPlayerIndex: numberAdapter,
})({
  currentPlayer: ({ playersAll, currentPlayerIndex }) => playersAll.at(currentPlayerIndex) || null,
  nextPlayerIndex: ({ currentPlayerIndex, playersIds }) => (currentPlayerIndex + 1) % playersIds.length,
})();
```

```ts
// Service
@Injectable({ providedIn: 'root' })
export class GameService {
  readonly #path = 'game';

  readonly #watchedAdapter = watch(this.#path, gameAdapter);

  // Sources
  readonly endCurrentRound$ = new Source<void>('[Game] End Current Round');
  readonly #nextPlayerIndex$ = this.endCurrentRound$.pipe(
    switchMap(() => this.#watchedAdapter.nextPlayerIndex$.pipe(take(1))),
    toSource('[Game] nextPlayerIndex$'),
  );

  readonly #store = adapt(initialState, {
    path: this.#path,
    adapter: gameAdapter,
    sources: {
      setCurrentPlayerIndex: this.#nextPlayerIndex$,
    },
  });

  readonly currentPlayer = toSignal<Player | null>(this.#store.currentPlayer$);
}
```
